aten/src/ATen/native/cpu/Depthwise.cpp

```diff
@@ -0,0 +1,298 @@
+                  2 * otw + 1 < args.out_cols
+              )) {
+        float32x4x4_t input_tile;
+        for (int row = 0; row < 4; ++row) {
+          input_tile.val[row] =
+              vld1q_f32(input + (ih + row) * args.in_cols + iw);
+        }
+
+        TILE;
+
+        for (size_t row = 0; row < 2; ++row) {
+          vst1_f32(
+              output + (oth * 2 + row) * args.out_cols + otw * 2,
+              vget_low_f32(input_tile.val[row]));
+        }
+      } else {
+        float block[4][4];
+        for (int row = 0; row < 4; ++row) {
+          for (int col = 0; col < 4; ++col) {
+            if (ih + row >= 0 && iw + col >= 0 && ih + row < args.in_rows &&
+                iw + col < args.in_cols) {
+              block[row][col] = input[(ih + row) * args.in_cols + iw + col];
+            } else {
+              block[row][col] = 0.0;
+            }
+          }
+        }
+
+        float32x4x4_t input_tile;
+        for (int row = 0; row < 4; ++row) {
+          input_tile.val[row] = vld1q_f32(&block[row][0]);
+        }
+
+        TILE;
+
+        float oblock[2][2];
+        for (int row = 0; row < 2; ++row) {
+          vst1_f32(&oblock[row][0], vget_low_f32(input_tile.val[row]));
+        }
+        for (int row = 0; row < 2; ++row) {
+          for (int col = 0; col < 2; ++col) {
+            if (2 * oth + row < args.out_rows &&
+                2 * otw + col < args.out_cols) {
+              output[(2 * oth + row) * args.out_cols + 2 * otw + col] =
+                  oblock[row][col];
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+#endif /* __ARM_NEON__ */
+
+Tensor _convolution_depthwise3x3_winograd(
+    const Tensor & input,
+    const Tensor & kernel,
+    const Tensor & bias_potentially_undefined,
+    const IntArrayRef stride,
+    const IntArrayRef padding,
+    const int groups)
+{
+    const IntArrayRef input_sizes = input.sizes();
+    const IntArrayRef kernel_sizes = kernel.sizes();
+
+    Tensor output = at::empty(
+      calculate_conv_output_size(input_sizes, kernel_sizes, stride, padding),
+      input.options());
+
+    const IntArrayRef output_sizes = output.sizes();
+
+    const Arguments args {
+        input_sizes[0],     // Input N
+        input_sizes[2],     // Input H
+        input_sizes[3],     // Input W
+        stride[0],          // Stride
+        padding[0],         // Padding Rows
+        padding[1],         // Padding Columns
+        output_sizes[2],    // Output H
+        output_sizes[3],    // Output W
+    };
+
+    const int input_hxw = args.in_rows * args.in_cols;
+    const int output_hxw = args.out_rows * args.out_cols;
+
+    const Tensor bias = bias_potentially_undefined.defined() ?
+                        bias_potentially_undefined :
+                        at::zeros({kernel_sizes[0]}, input.options());
+
+    const auto f = [&](const int n, const int g) {
+      convolution_depthwise3x3_winograd_impl(
```

Comment:
so how does this code compile outside of NEON? you'd probably need to propagate the setting all the way up to Convolution.cpp. If you don't want to put #ifdef there - you can just create bool constexpr function/constant which you define here based on whether implementation is available or not

Reply:
Thanks for the suggestion.